Implement relaxed laneselect SIMD instructions

wpaulino opened the pull request:
#### 9affbf98e0099fe8f126bd9edba69f559d7fc33f
<pre>
Implement relaxed laneselect SIMD instructions
<a href="https://bugs.webkit.org/show_bug.cgi?id=281581">https://bugs.webkit.org/show_bug.cgi?id=281581</a>

Reviewed by Yusuke Suzuki.

The specification[1] notes that these instructions should behave as a
bitwise select SIMD instruction when the mask has either _all_ its bits
set or unset. When the mask is of any other value, it allows for
implementations to choose the behavior of the relaxed instruction based
on the backing architecture to be executed on.

The relaxed laneselect instructions are implemented for ARM64 as the
bitwise select (bsl) SIMD instruction since they are already natively
supported by the architecture.

For x86_64, CPUs without AVX support do not have an equivalent bsl
instruction. CPUs with SSE4.1 support feature an alternative set of
instructions (pblend) in which only the top bit of each lane within the
mask is examined. JSC requires AVX support for SIMD instructions, so
this is not applicable.

CPUs with AVX support have a similar instruction to ARM64&apos;s bsl, known
as vpblend. Note that this instruction is not currently used, and it is
instead emulated as v128.or(v128.and(v1, c), v128.and(v2, v128.not(c))).
This optimization can be deferred to future work. Nonetheless, we also
implement the relaxed laneselect instructions as a bitwise select for
x86_64 CPUs with AVX support.

[1] <a href="https://github.com/WebAssembly/relaxed-simd/blob/main/proposals/relaxed-simd/Overview.md#relaxed-laneselect">https://github.com/WebAssembly/relaxed-simd/blob/main/proposals/relaxed-simd/Overview.md#relaxed-laneselect</a>

* JSTests/wasm/stress/simd-const-relaxed-lane-select.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.result.i32.i8x16.relaxed_laneselect.v128.const.i8x16.0.127.0.0.0.0.0.0.0.0.0.0.0.0.0.0.v128.const.i8x16.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.v128.const.i8x16.0.7.0.0.0.0.0.0.0.0.0.0.0.0.0.0.return.i8x16.extract_lane_u.1.func.export.string_appeared_here.result.i32.i16x8.relaxed_laneselect.v128.const.i16x8.0.32767.0.0.0.0.0.0.v128.const.i16x8.0.0.0.0.0.0.0.0.v128.const.i16x8.0.7.0.0.0.0.0.0.return.i16x8.extract_lane_u.1.func.export.string_appeared_here.result.i32.i32x4.relaxed_laneselect.v128.const.i32x4.0.2147483647.0.0.v128.const.i32x4.0.0.0.0.v128.const.i32x4.0.7.0.0.return.i32x4.extract_lane.1.func.export.string_appeared_here.result.i64.i64x2.relaxed_laneselect.v128.const.i64x2.0.9223372036854775807.v128.const.i64x2.0.0.v128.const.i64x2.0.7.return.i64x2.extract_lane.1.async test):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSIMDBitwiseSelect):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):

Canonical link: <a href="https://commits.webkit.org/286162@main">https://commits.webkit.org/286162@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21ebfd588d61aaacfdf6c7971d27b860f9d18ecf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72272 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51693 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76442 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23481 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59497 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23303 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56961 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15462 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46827 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62235 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37393 "Found 1 new API test failure: /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43483 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; 4 new passes 5 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19711 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21831 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65401 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20070 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78118 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71526 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16513 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19221 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65429 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16560 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62259 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64694 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12920 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6554 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93307 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11574 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47491 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20536 "Found 3 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/verbose-failure-dont-graph-dump-availability-already-freed.js.default, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48560 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49848 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48303 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
Can you add the explanation of this patch in a commit message? the strategy, the difference from the non relaxed lane select, and how it is implemented. (also link to the spec)

Constellation reviewed (approved):
r=me

wpaulino committed:
Implement relaxed laneselect SIMD instructions
https://bugs.webkit.org/show_bug.cgi?id=281581

Reviewed by Yusuke Suzuki.

The specification[1] notes that these instructions should behave as a
bitwise select SIMD instruction when the mask has either _all_ its bits
set or unset. When the mask is of any other value, it allows for
implementations to choose the behavior of the relaxed instruction based
on the backing architecture to be executed on.

The relaxed laneselect instructions are implemented for ARM64 as the
bitwise select (bsl) SIMD instruction since they are already natively
supported by the architecture.

For x86_64, CPUs without AVX support do not have an equivalent bsl
instruction. CPUs with SSE4.1 support feature an alternative set of
instructions (pblend) in which only the top bit of each lane within the
mask is examined. JSC requires AVX support for SIMD instructions, so
this is not applicable.

CPUs with AVX support have a similar instruction to ARM64's bsl, known
as vpblend. Note that this instruction is not currently used, and it is
instead emulated as v128.or(v128.and(v1, c), v128.and(v2, v128.not(c))).
This optimization can be deferred to future work. Nonetheless, we also
implement the relaxed laneselect instructions as a bitwise select for
x86_64 CPUs with AVX support.

[1] https://github.com/WebAssembly/relaxed-simd/blob/main/proposals/relaxed-simd/Overview.md#relaxed-laneselect

* JSTests/wasm/stress/simd-const-relaxed-lane-select.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.result.i32.i8x16.relaxed_laneselect.v128.const.i8x16.0.127.0.0.0.0.0.0.0.0.0.0.0.0.0.0.v128.const.i8x16.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.v128.const.i8x16.0.7.0.0.0.0.0.0.0.0.0.0.0.0.0.0.return.i8x16.extract_lane_u.1.func.export.string_appeared_here.result.i32.i16x8.relaxed_laneselect.v128.const.i16x8.0.32767.0.0.0.0.0.0.v128.const.i16x8.0.0.0.0.0.0.0.0.v128.const.i16x8.0.7.0.0.0.0.0.0.return.i16x8.extract_lane_u.1.func.export.string_appeared_here.result.i32.i32x4.relaxed_laneselect.v128.const.i32x4.0.2147483647.0.0.v128.const.i32x4.0.0.0.0.v128.const.i32x4.0.7.0.0.return.i32x4.extract_lane.1.func.export.string_appeared_here.result.i64.i64x2.relaxed_laneselect.v128.const.i64x2.0.9223372036854775807.v128.const.i64x2.0.0.v128.const.i64x2.0.7.return.i64x2.extract_lane.1.async test):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSIMDBitwiseSelect):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):

Canonical link: https://commits.webkit.org/286162@main
